server/src/main/java/org/opensearch/common/cache/stats/MultiDimensionCacheStats.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.stats;
+
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * A CacheStats object supporting aggregation over multiple different dimensions.
+ * Also keeps track of a tier dimension, which is the same for all values in the stats object.
+ * Does not allow changes to the stats.
+ */
+public class MultiDimensionCacheStats implements CacheStats {
+
+    // The value of the tier dimension for entries in this Stats object. This is handled separately for efficiency,
+    // as it always has the same value for every entry in the stats object.
+    // Package-private for testing.
+    final String tierDimensionValue;
```

Comment:
We're allowing the user to set level=tier for the new API. When the CacheService gets stats from the IndicesRequestCache, it doesn't know whether the stats is from a tiered cache or something else, so I thought it would make the most sense if the stats object itself could correctly aggregate by dimension. Otherwise we'd have to put separate cases in CacheService for different IRC implementations, right?

Reply:
@msfroh I think we need to go back to cache stats API response to argue this.

As @peteralfonsi mentioned here, we are thinking to allow user set ```level=tier``` in stats request API. So I see below option:

**if Cache is TieredCache:**

`GET /_nodes/stats/caches/request_cache?level=indices,tier&pretty`

```
// Only including hit count for as an example.
"caches": {
   "request_cache": { 
        "hit_count" : 4,
        "indices": {
            "index1": {
                "tier": {
                    "onheap": {
                        "hit_count" : 2
                    },
                    "disk": {
                        "hit_count" : 2
                    }
                }
            }
        }
    }
}
```


**if Cache is OnHeapCache:**

`GET /_nodes/stats/caches/request_cache?level=indices,tier&pretty`

If the user **tier** as a dimension here, but cache is onHeap, then we still include "tier" in the stats API to keep the response consistent, right?

```
"caches": {
   "request_cache": { 
        "hit_count" : 4,
        "indices": {
            "index1": {
                "tier": {
                    "onheap": {
                        "hit_count" : 4
                    }
                }
            }
        }
    }
}
```

Reply:
In theory, if the stats are held with all the dimensions, you could aggregate on any dimensions supported by the cache, without needing to know what they are.

That is, the dimension names are just strings provided by the cache implementation. The core doesn't need to know what those dimension names are in order to aggregate on them.

Reply:
We could store the tier dimension within ICacheKey itself. The tradeoff would be an extra ~20B for each cache key. Would this be better? 

We were thinking we could do the same thing for the indices dimension. This would spare us from having to use the map from indices -> shards during aggregation in the CacheService. CacheStats would then know everything it needs to to aggregate by shard, tier, or indices, so we could have CacheStats itself provide its own XContent, and we could remove the getStatsByDimensions() API.

Reply:
>In theory, if the stats are held with all the dimensions, you could aggregate on any dimensions supported by the cache

Agree on that. We were thinking same. We were just torn on the "tier" dimension as it comes by default with each cache implementation.
So for IndicesRequestCache, we will have below dimensions(without tier). These can be passed as part of ICacheKey.
```
"shards", "indices", "indices#shards" (to support the combination of indices/shards)
```

Now to maintain "tier" dimension we can append that in above dimensions, now we have 3X dimensions.

```
"shards", "shards#tier",
"indices", "indices#tier",
"indices#shards", "indices#shards#tier"
```
Pros:
Easy to write generic code.
Cons:
Takes extra space in map per key to have these redundant tier dimension values As it isn't necessary to handle considering it comes by default.

Reply:
You don't need to store all of the aggregations, though, since you can aggregate when the stats are requested, right? 

If you just store stats for `index#shard#tier`, then each `index` value would just sum across all `shard` and `tier` values for the given index.

Reply:
Yes, we can store just `index#shard#tier` in the CacheStats. We would still have to store indices, tier, and shard dimensions within each ICacheKey, instead of just shard. That would require an extra 2 dimensions stored in each key, which would take up extra space in both caches (~4% of each key-value pair), but also let us have each CacheStats produce its own XContent. 

Let us know if you think this is a good approach and I will start making those changes!

Reply:
The other option is to let the cache implementation own its own stats.

While the on heap cache and the EHDiskCache can each track stats for at the shard level, the tiered cache knows which underlying cache is on which tier (which again, is the tiered cache's responsibility, not the underlying cache's responsibility). 

Then the tiered cache can delegate to the individual caches to ask for their stats and emit them per tier.

Reply:
Right, but the issue comes if someone is using the IRC with the OpenSearchOnHeap cache, no tiered cache, and requests to aggregate with level=tier.